libuavcan/src/transport/can_acceptance_filter_configurator.cpp

```diff
@@ -0,0 +1,256 @@
+    while (p1)
+    {
+        CanFilterConfig cfg;
+        cfg.id = DefaultFilterServiceRequestID;
+        cfg.id |= static_cast<uint32_t>(p1->getDataTypeDescriptor().getID().get()) << 17;
+        cfg.mask = DefaultFilterServiceRequestMask;
+
+        multiset_configs_.emplace(cfg);
+        if (multiset_configs_.getByIndex(i) == NULL)
+        {
+            UAVCAN_ASSERT(0);
+        }
+
+        i++;
+        p1 = p1->getNextListNode();
+    }
+
+    if (multiset_configs_.getSize() == 0)
+    {
+        return -ErrDriver;
+    }
+
+    for (uint16_t i = 0; i < multiset_configs_.getSize(); i++)
+    {
+        UAVCAN_TRACE("CanAcceptanceFilterConfigurator::fillArray()", "map.ID [%u] = %d", i,
+                     multiset_configs_.getByIndex(i)->id);
+        UAVCAN_TRACE("CanAcceptanceFilterConfigurator::fillArray()", "map.MK [%u] = %d", i,
+                     multiset_configs_.getByIndex(i)->mask);
+    }
+
+    return 0;
+}
+
+int16_t CanAcceptanceFilterConfigurator::computeConfiguration()
+{
+    uint16_t i_rank = 0, j_rank = 0;
+    const uint16_t acceptance_filters_number = getNumFilters();
+
+    while (acceptance_filters_number < multiset_configs_.getSize())
+    {
+        UAVCAN_ASSERT(multiset_configs_.getSize() != 0);
+
+        uint8_t best_rank = 0;
+        const uint16_t map_array_size = static_cast<uint16_t>(multiset_configs_.getSize());
+
+        for (uint16_t i_ind = 0; i_ind < map_array_size - 1; i_ind++)
+        {
+            for (uint16_t j_ind = static_cast<uint16_t>(i_ind + 1); j_ind < map_array_size; j_ind++)
+            {
+                CanFilterConfig temp_array = mergeFilters(*multiset_configs_.getByIndex(i_ind),
+                                                          *multiset_configs_.getByIndex(j_ind));
+                uint8_t rank = countBits(temp_array.mask);
+                if (rank > best_rank)
+                {
+                    best_rank = rank;
+                    i_rank = i_ind;
+                    j_rank = j_ind;
+                }
+            }
+        }
+
+        *multiset_configs_.getByIndex(j_rank) = mergeFilters(*multiset_configs_.getByIndex(i_rank),
+                                                             *multiset_configs_.getByIndex(j_rank));
+        CanFilterConfig temp_key = *multiset_configs_.getByIndex(i_rank);
+
+        if ((multiset_configs_.getByIndex(i_rank)->id == multiset_configs_.getByIndex(j_rank)->id) &&
+            (multiset_configs_.getByIndex(i_rank)->mask == multiset_configs_.getByIndex(j_rank)->mask))
+        {
+            multiset_configs_.removeAll(temp_key);             // If i_rank and j_rank elements are the same, we need to add it
+            multiset_configs_.emplace(temp_key);               //another time, to preclude removal of both elements
+        }
+        else
+        {
+            multiset_configs_.removeAll(temp_key);
+        }
```

Comment:
It deletes only the first occurrence of an element with the same value (and also it may or may not shuffle some or all of other elements). Unless I'm terribly wrong, the algorithm doesn't care about which particular copy gets removed, so that should be safe.